### PR TITLE
Add (Un)Blind Samples Button on Box Detail

### PR DESCRIPTION
**Pull request by omelao**
Just for Owners/Creators, if a box is transfered it will not show the button.

image

image

### CODE REVIEW

File: app/models/box.rb [hunk 1]
```diff
@@ -94,4 +94,12 @@ def blind_attribute?(attr_name)
   def transferred?
     box_transfers.exists?
   end
+
+  def unblind!
+    self.update(blinded: false)
```
Comment (ysbaddaden) [inline app/models/box.rb:99]: `self` is implicit in Ruby, you don't have to specify it (same in `#blind!`). You may also want to use [#update_columns](https://api.rubyonrails.org/v6.1/classes/ActiveRecord/Persistence.html#method-i-update_columns) which would be a bit faster (no need to re-validate the box, we only change one value):

```suggestion
    update_columns(blinded: false)
```

Same for `#blind!`

File: app/controllers/boxes_controller.rb [hunk 2]
```diff
@@ -17,6 +17,7 @@ def index
   def show
     return unless authorize_resource(@box, READ_BOX)
     @can_delete = has_access?(@box, DELETE_BOX)
+    @can_blind = @box.box_transfers.count == 0
```
Comment (ysbaddaden) [inline app/controllers/boxes_controller.rb:20]: We have a method for that (it's a little faster as it won't even run a COUNT in SQL):

```suggestion
    @can_blind = !@box.transferred?
```

File: app/controllers/boxes_controller.rb [hunk 3]
```diff
@@ -91,6 +92,21 @@ def bulk_destroy
     redirect_to boxes_path, notice: "Boxes were successfully deleted."
   end
 
+  def unblind
+    return unless authorize_resource(@box, UPDATE_BOX)
```
Comment (ysbaddaden) [inline app/controllers/boxes_controller.rb:96]: Sadly, this won't do. Once the box is added to a transfer we won't show the blind/unblind button, but once the transfer is confirmed, the receiver institution will own the box, and thus be authorized to blind/unblind by forging the request.

A basic `return head :forbidden if @box.transferred?` should work for now.

**Review by ysbaddaden — CHANGES_REQUESTED** [review]
Nice! Just a few readability enhancements, but most importantly the authorization to improve.

**Comment by ysbaddaden** [comment]
@omelao Can I bother you to add controller specs for the `boxes#blind` and `boxes#unblind` actions, to ensure that they work as expected?

- context: before transfer
  - owner institution can blind box
  - owner institution can unblind box

- context: after transfer
  - owner institution can't blind box
  - owner institution can't unblind box
  - sender institution can unblind box after transfer

Those specs are important to ensure that we don't inadvertently break the authorizations later!

**Comment by omelao** [comment]
Yeah. Sure.

**Comment by omelao** [comment]
@ysbaddaden done

**Pull request merged**